### PR TITLE
Import xlsx

### PR DESCRIPTION
**Pull request by junajan**
This PR allows users to export data into excel XLSX format.
As it was in export-xlsx branch, I wanted to keep backward compatibility and do a refactor how the data are loaded but not so big changes because of small time capacity for this task and because it is getting quite urgent now.

Also note that this PR goes to export-xlsx branch so we can see only changes in import script

### CODE REVIEW

File: src/coffee/import.coffee [hunk 1]
```diff
@@ -38,26 +39,34 @@ class Import
...
     @publishProducts = false
     @continueOnProblems = options.continueOnProblems
     @allowRemovalOfVariants = false
-    @updatesOnly = false
-    @dryRun = false
     @blackListedCustomAttributesForUpdate = []
     @customAttributeNameToMatch = undefined
     @matchBy = CONS.HEADER_ID
     @options = options
+#    @initializeObjects()
```
Comment (Siilwyn) [inline src/coffee/import.coffee]: What's up with this?

File: src/coffee/run.coffee [hunk 2]
```diff
@@ -103,8 +103,11 @@ module.exports = class
     program
       .command 'import'
       .description 'Import your products from CSV into your SPHERE.IO project.'
-      .option '-c, --csv <file>', 'CSV file containing products to import'
-      .option '-z, --zip <file>', 'ZIP archive containing multiple product files to import'
+      .option '-c, --csv <file>', 'CSV file containing products to import (alias for "in" parameter)'
+      # add alias for csv parameter and "-i" is taken so use "-f" parameter
+      .option '-f, --in <file>', 'File containing products to import'
+      .option '-z, --zip', 'Input file is archived'
+      .option '-x, --xlsx', 'Import from XLSX format'
```
Comment (Siilwyn) [inline src/coffee/run.coffee]: Would using `-f, --format <csv or xlsx>` be an idea here? That way there is no chance for conflicting flags.

File: src/coffee/import.coffee (hunk 1 above)
Comment (junajan) [inline src/coffee/import.coffee]: Initialization is now done when doing import - alias here:
https://github.com/sphereio/sphere-node-product-csv-sync/pull/137/files/9369cd729e709c08947c0fcd48fdb46f9bc1a363#diff-7ed9e504a5a8e3612516e56fc3dfd602R86

Comment (junajan) [inline src/coffee/import.coffee]: I will remove this commented code because it is not needed anymore

File: src/coffee/run.coffee (hunk 2 above)
Comment (Siilwyn) [inline src/coffee/run.coffee]: Oh sorry, I forgot about putting it in the next PR. Sounds like a good idea! 👍

File: src/coffee/io/reader.coffee [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  read: (file) =>
+    # read from file or from stdin?
+    if file
+      debugLog "READER::stream file %s", file
+      @inputStream = fs.createReadStream file
+    else
+      debugLog "READER::stream stdin"
+      @inputStream = process.stdin
+
+    if @options.format == 'xlsx'
```
Comment (hisabimbola) [inline src/coffee/io/reader.coffee]: I guess we agreed to use `exportFormat`

File: src/coffee/validator.coffee [hunk 4]
```diff
@@ -29,18 +29,23 @@ class Validator
       delimiter: options.csvDelimiter or ','
       quote: options.csvQuote or '"'
 
-  parse: (csvString) ->
+  parse: (csvString, delimiter ) ->
```
Comment (hisabimbola) [inline src/coffee/validator.coffee]: why you parsing `delimiter` here. 

It's not being used anywhere in this method.

File: src/spec/integration/importArchive.spec.coffee [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+Excel = require 'exceljs'
+cuid = require 'cuid'
+path = require 'path'
+tmp = require 'tmp'
+fs = Promise.promisifyAll require('fs')
+# will clean temporary files even when an uncaught exception occurs
+tmp.setGracefulCleanup()
+
+createImporter = (format) ->
+  Config.importFormat = format || "csv"
```
Comment (hisabimbola) [inline src/spec/integration/importArchive.spec.coffee]: why are we updating the `Config` object in the test?

Not sure what's it's doing here, but if you want to update the properties in it, you can clone it and alter.

File: src/spec/integration/importArchive.spec.coffee [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      headers = []
+      for i of items
+        headers.push {
+          header: items[i]
+        }
+      worksheet.columns = headers
+
+  workbook.xlsx.writeFile(filePath)
+
+CHANNEL_KEY = 'retailerA'
```
Comment (hisabimbola) [inline src/spec/integration/importArchive.spec.coffee]: Best practice for constants or variables to be declared at the top of the file/function.

File: src/coffee/import.coffee [hunk 7]
```diff
@@ -103,27 +122,47 @@ class Import
...
+  _unarchiveProducts: (archivePath) ->
     tempDir = tmp.dirSync({ unsafeCleanup: true })
-    console.log "Importing archive #{archivePath}"
+    console.log "Unarchiving file #{archivePath}"
 
     extractArchive(archivePath, {dir: tempDir.name})
     .then =>
-      console.log "Loading files from", tempDir.name
-      filePaths = walkSync tempDir.name, { globs: ['**/*.csv'] }
+      filePredicate = '**/*.'+@options.importFormat
```
Comment (hisabimbola) [inline src/coffee/import.coffee]: Can you coffeescript templating string here.

File: src/coffee/io/reader.coffee (hunk 3 above)
Comment (hisabimbola) [inline src/coffee/io/reader.coffee]: I think it will be `importFormat` here.

File: src/coffee/io/reader.coffee [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        resolve(rows)
+
+  _readCsv: (stream) =>
+    new Promise (resolve, reject) =>
+      buffers = []
+
+      # stream whole file to buffer because we need to decode it first from buffer
+      # - iconv-lite does not support string to string decoding
+      stream.on 'data', (buffer) ->
+        buffers.push buffer
```
Comment (hisabimbola) [inline src/coffee/io/reader.coffee]: is there anyway to do this without storing all the data in memory?

Let's discuss.

Comment (junajan) [inline src/coffee/io/reader.coffee]: Now it works as before (it loads the whole file before starting an import). I think that the main reason here is that we need all productVariants before updating/importing product. It should be possible to stream file lines and emit `data` event when we have all productRows (masterVariant and variants) loaded - once we load new masterVariant, we know that the loading of previous product has finished so we can process him further.

File: src/coffee/import.coffee (hunk 1 above)
Comment (junajan) [inline src/coffee/import.coffee]: updated here 50e69a4327638c5111d5ca360654e160b01c37aa

File: src/coffee/io/reader.coffee (hunk 3 above)
Comment (junajan) [inline src/coffee/io/reader.coffee]: updated here bb53bd6

File: src/coffee/validator.coffee (hunk 4 above)
Comment (junajan) [inline src/coffee/validator.coffee]: removed here bb53bd6

File: src/spec/integration/importArchive.spec.coffee (hunk 6 above)
Comment (junajan) [inline src/spec/integration/importArchive.spec.coffee]: moved to the top here bb53bd6

File: src/coffee/import.coffee (hunk 7 above)
Comment (junajan) [inline src/coffee/import.coffee]: updated here bb53bd6

**Comment by junajan** [comment]
Hello @hisabimbola - I fixed tests and updated code based on your PR review..

**Pull request merged**